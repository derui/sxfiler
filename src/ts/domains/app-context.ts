import { UIContext } from "@/types/ui-context";

export type AppContext = {
  readonly current: UIContext;
  readonly subContexts: UIContext[];
};

/**
   create AppContext
 */
export function createAppContext({
  current,
  subContexts,
}: {
  current: UIContext;
  subContexts?: UIContext[];
}): AppContext {
  return {
    current,
    subContexts: Array.from(subContexts || []),
  };
}

/**
   change context
 */
export function changeCurrent(context: UIContext) {
  return (state: AppContext) =>
    createAppContext({
      current: context,
      subContexts: state.subContexts,
    });
}

/**
   Add a context to AppContext
 */
export function addSubContext(context: UIContext) {
  return (state: AppContext) => {
    const tmpSet = new Set(state.subContexts);
    tmpSet.add(context);

    return createAppContext({
      current: state.current,
      subContexts: Array.from(tmpSet.values()),
    });
  };
}

/**
   remove the context from AppContext
 */
export function removeSubContext(context: UIContext) {
  return (state: AppContext) => {
    const tmpSet = new Set(state.subContexts);
    tmpSet.delete(context);

    return createAppContext({
      current: state.current,
      subContexts: Array.from(tmpSet.values()),
    });
  };
}
