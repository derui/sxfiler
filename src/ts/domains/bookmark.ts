/**
   Type of bookmark
 */
export type Bookmark = {
  readonly id: string;
  readonly path: string;
  readonly order: number;
};

type FactoryArg = Pick<Bookmark, keyof Bookmark>;

/**
 * Create bookmark from factory argument
 */
export const createBookmark = function createBookmark({ id, path, order }: FactoryArg): Bookmark {
  return {
    id,
    path,
    order,
  };
};
