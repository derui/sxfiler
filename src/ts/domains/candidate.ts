export type CandidateObject = {
  readonly id: string;
  readonly value: string;
  readonly start: number;
  readonly length: number;
};

export type Candidate = CandidateObject & {
  /**
   * Get splitted candidate as 3 pieces by matched string that
   * has the one of before start of match, matched, and after matched.
   */
  splitByInput(): [string, string, string];

  /**
   * Get plain object.
   */
  plain(): CandidateObject;
};

type FactoryArg = Pick<CandidateObject, "id" | "value"> & Partial<Pick<CandidateObject, "start" | "length">>;

/**
 * Create candidate from candidate object.
 */
export const createCandidate = ({ id, value, start = 0, length = 0 }: FactoryArg): Candidate => {
  return {
    id,
    value,
    start: Math.max(0, Math.min(value.length - 1, start)),
    length: Math.max(0, Math.min(value.length, length)),
    splitByInput() {
      const beforeMatched = value.substring(0, start);
      const matched = value.substring(start, start + length);
      const afterMatched = value.substring(start + length);

      return [beforeMatched, matched, afterMatched];
    },

    plain() {
      return { id: this.id, value: this.value, start: this.start, length: this.length };
    },
  };
};
