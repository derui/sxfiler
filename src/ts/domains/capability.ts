// Capability with method
export type Capability = CapabilityObject & {
  /** Return new capability that allows to write */
  allowToWrite(): Capability;

  /** Return new capability that allows to read */
  allowToRead(): Capability;

  /** Return new capability that allows to execute */
  allowToExecute(): Capability;

  /** Return new capability that disallows to write */
  disallowToWrite(): Capability;

  /** Return new capability that disallows to read */
  disallowToRead(): Capability;

  /** Return new capability that disallows to execute */
  disallowToExecute(): Capability;

  /** Return plain CapabilityObject */
  plain(): CapabilityObject;
};

// Capability of a target
export type CapabilityObject = {
  readonly writable: boolean;
  readonly readable: boolean;
  readonly executable: boolean;
};

type CreateCapabilityArg = {
  writable: boolean;
  readable: boolean;
  executable: boolean;
};

function plain(this: Capability): CapabilityObject {
  return { writable: this.writable, readable: this.readable, executable: this.executable };
}

/**
   Create capability from argument
 */
export const createCapability = ({ writable, readable, executable }: CreateCapabilityArg): Capability => {
  return {
    writable,
    readable,
    executable,

    plain,

    /** Return new capability that allows to write */
    allowToWrite() {
      return { ...this, writable: true };
    },

    /** Return new capability that allows to read */
    allowToRead() {
      return { ...this, readable: true };
    },

    /** Return new capability that allows to execute */
    allowToExecute() {
      return { ...this, executable: true };
    },

    /** Return new capability that disallows to write */
    disallowToWrite() {
      return { ...this, writable: false };
    },

    /** Return new capability that disallows to read */
    disallowToRead() {
      return { ...this, readable: false };
    },

    /** Return new capability that disallows to execute */
    disallowToExecute() {
      return { ...this, executable: false };
    },
  };
};

/**
   Return new capability that have not any work.
 */
export const emptyCapability = (): Capability =>
  createCapability({ writable: false, readable: false, executable: false });

/**
   Return new capability that have all capability on all roles.
 */
export const fullCapability = (): Capability => createCapability({ writable: true, readable: true, executable: true });
