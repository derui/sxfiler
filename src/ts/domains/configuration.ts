// value object for configuration of application

enum SortType {
  Name,
  Date,
  Timestamp,
}

export class Configuration {
  public readonly defaultSortOrder: SortType = SortType.Name;
}
