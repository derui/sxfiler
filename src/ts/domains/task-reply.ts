type OverwritePayload = {
  readonly kind: "overwrite";
};

type RenamePayload = {
  readonly kind: "rename";
  readonly newName: string;
};

type PayloadObject = OverwritePayload | RenamePayload;
export type Payload = PayloadObject;

type ReplyObject = {
  readonly reply: Payload;
  readonly taskId: string;
};
export type Reply = ReplyObject;

// create payload to overwrite a node
export const createOverwritePayload = (): Payload => {
  return {
    kind: "overwrite",
  };
};

// create payload to rename a node
export const createRenamePayload = (newName: string): Payload => {
  return {
    kind: "rename",
    newName,
  };
};

// create object for reply
export const createReply = (taskId: string, payload: Payload): Reply => {
  return { taskId, reply: payload };
};
