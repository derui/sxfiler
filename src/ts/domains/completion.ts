import { Candidate, CandidateObject, createCandidate } from "./candidate";
import { between } from "@/utils";

export type CompletionObject = {
  // current key map
  readonly cursor: number;
  readonly candidates: CandidateObject[];
};

export type Completion = {
  // current key map
  readonly cursor: number;
  readonly candidates: Candidate[];

  /**
     replace candidates with new candidates
   */
  replaceCandidates(candidates: Candidate[]): Completion;

  /**
     move cursor to amount value
   */
  moveCursor(amount: number): Completion;

  /**
     return plain object
   */
  plain(): CompletionObject;
};

function replaceCandidates(this: Completion, candidates: CandidateObject[]): Completion {
  return { ...this, candidates: candidates.map(createCandidate) };
}

function moveCursor(this: Completion, amount: number): Completion {
  return { ...this, cursor: between(this.cursor + amount, this.candidates.length - 1) };
}

function plain(this: Completion): CompletionObject {
  return { cursor: this.cursor, candidates: this.candidates.map(v => v.plain()) };
}

/** return empty state */
export const createCompletion = ({ cursor, candidates }: Partial<CompletionObject>): Completion => {
  if (cursor && candidates) {
    cursor = between(cursor, candidates.length);
  } else if (candidates) {
    cursor = 0;
  } else {
    cursor = 0;
    candidates = [];
  }
  return { cursor, candidates: candidates.map(createCandidate), replaceCandidates, plain, moveCursor };
};
