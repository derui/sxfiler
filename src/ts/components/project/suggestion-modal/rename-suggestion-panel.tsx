import * as React from "react";
import { Suggestion, SuggestionKind } from "../../../domains/task-suggestion";
import * as Element from "../../ui/element/element";
import { ReplyPayload, createRenamePayload } from "../../../domains/task-reply";

const style: ClassNames = require("./rename-suggestion-panel.module.scss");

type ClassNames = {
  root: string;
  text: string;
  label: string;
  input: string;
  labelContainer: string;
};

type Handler = (payload: ReplyPayload) => void;

export type Props = {
  selected: boolean;
  suggestion: Suggestion;
  onReply: Handler;
  nodeName: string;
};

type State = {
  newName: string;
};

export class Component extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = { newName: props.nodeName };
  }

  private handleChange = () => (ev: React.ChangeEvent) => {
    this.setState({ newName: ev.target.nodeValue || "" });
  };

  private handleKeyDown = () => (ev: React.KeyboardEvent) => {
    if (!this.props.selected || this.state.newName === "") {
      return;
    }

    if (ev.key === "Enter") {
      ev.stopPropagation();
      ev.preventDefault();
      this.props.onReply(createRenamePayload(this.state.newName));
    }
  };

  render() {
    const { selected, suggestion } = this.props;
    if (suggestion.kind !== SuggestionKind.Rename) {
      return null;
    }

    return (
      <Element.Component className={style.root} onKeyDown={this.handleKeyDown()}>
        <p className={style.text} aria-selected={selected}>{`Rename`}</p>
        <label className={style.labelContainer} aria-selected={selected}>
          <span className={style.label}>New Name</span>
          <input className={style.input} type="text" value={this.state.newName} onChange={this.handleChange()} />
        </label>
      </Element.Component>
    );
  }
}
