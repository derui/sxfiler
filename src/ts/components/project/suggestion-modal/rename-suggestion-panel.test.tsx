import * as React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import { Component } from "./rename-suggestion-panel";
import { createRenamePayload, ReplyPayload } from "../../../domains/task-reply";

describe("Project", () => {
  describe("Rename suggestion", () => {
    it("render with rename suggestion", () => {
      const handler = () => {};
      const tree = renderer.create(<Component selected={false} onUpdated={handler} nodeName={"node"} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("change style when selected", () => {
      const handler = () => {};
      const tree = renderer.create(<Component selected={true} onUpdated={handler} nodeName={"node"} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("do not render with other suggestion", () => {
      const handler = () => {};
      const tree = renderer.create(<Component selected={false} onUpdated={handler} nodeName={"node"} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("call handler when the new name in the component is changed", done => {
      const handler = (reply: ReplyPayload) => {
        expect(reply).toEqual(createRenamePayload("foo"));
        done();
      };
      const wrapper = mount(<Component selected={true} onUpdated={handler} nodeName="bar" />);
      wrapper.find("input").simulate("change", {
        target: {
          nodeValue: "foo",
        },
      });
    });
  });
});
