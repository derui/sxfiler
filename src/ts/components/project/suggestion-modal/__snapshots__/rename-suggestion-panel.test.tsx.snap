// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Rename suggestion change style when selected 1`] = `
<div
  aria-selected={true}
  className="sc-bdVaJa hTXxV"
>
  <p>
    Rename
  </p>
  <label>
    <input
      className="sc-bwzfXH fhlOTR"
      onChange={[Function]}
      type="text"
      value="node"
    />
  </label>
</div>
`;

exports[`Project Rename suggestion do not render with other suggestion 1`] = `
<div
  aria-selected={false}
  className="sc-bdVaJa hTXxV"
>
  <p>
    Rename
  </p>
  <label>
    <input
      className="sc-bwzfXH fhlOTR"
      onChange={[Function]}
      type="text"
      value="node"
    />
  </label>
</div>
`;

exports[`Project Rename suggestion render with rename suggestion 1`] = `
<div
  aria-selected={false}
  className="sc-bdVaJa hTXxV"
>
  <p>
    Rename
  </p>
  <label>
    <input
      className="sc-bwzfXH fhlOTR"
      onChange={[Function]}
      type="text"
      value="node"
    />
  </label>
</div>
`;
